substrate/network-libp2p/src/service.rs

```diff
@@ -0,0 +1,1106 @@
+			out.info
+				.and_then(move |info|
+					if info.info.public_key.into_peer_id() == expected_peer_id {
+						Ok(socket)
+					} else {
+						debug!(target: "sub-libp2p", "Public key mismatch for \
+							node {:?} with  proto {:?}", expected_peer_id, proto_id);
+						trace!(target: "sub-libp2p", "Removing addr {} for {:?}",
+							original_addr, expected_peer_id);
+						shared.network_state.set_invalid_kad_address(&expected_peer_id, &original_addr);
+						Err(IoErrorKind::InvalidData.into())		// TODO: correct err
+					}
+				)
+				.and_then(move |socket|
+					upgrade::apply(socket, proto, endpoint, client_addr)
+				)
+		})
+		.and_then(move |out, endpoint, client_addr|
+			future::ok(((FinalUpgrade::Custom(out), endpoint), client_addr))
+		);
+
+	swarm_controller.dial(addr, with_proto)
+		.map_err(|_| IoError::new(IoErrorKind::Other, "multiaddr not supported"))
+}
+
+/// Obtain a Kademlia connection to the given peer.
+fn obtain_kad_connection<T, To, St, C>(shared: Arc<Shared>,
+	peer_id: PeerstorePeerId, transport: T, swarm_controller: SwarmController<St>)
+	-> impl Future<Item = KadConnecController, Error = IoError>
+	where T: MuxedTransport<Output =  TransportOutput<To>> + Clone + 'static,
+		T::MultiaddrFuture: 'static,
+		To: AsyncRead + AsyncWrite + 'static,
+		St: MuxedTransport<Output = (FinalUpgrade<C>, Endpoint)> + Clone + 'static,
+		C: 'static {
+	let kad_upgrade = shared.kad_upgrade.clone();
+	let final_future = shared.network_state
+		.obtain_kad_connection(peer_id.clone(), move || {
+			let addr: Multiaddr = AddrComponent::P2P(peer_id.clone().into_bytes()).into();
+			trace!(target: "sub-libp2p", "Opening new kademlia connection to {}", addr);
+			let (tx, rx) = oneshot::channel();
+			let tx = Arc::new(Mutex::new(Some(tx)));
+			swarm_controller.dial(addr, 
+				transport
+					.and_then(move |out, endpoint, client_addr|
+						upgrade::apply(out.socket, kad_upgrade.clone(),
+							endpoint, client_addr)
+					)
+					.map(move |(kad_ctrl, stream), _| {
+						if let Some(tx) = tx.lock().take() {
+							let _ = tx.send(kad_ctrl.clone());
+						}
+						(FinalUpgrade::Kad((kad_ctrl, stream)), Endpoint::Dialer)
+					})
+				)
+				.expect("cannot dial");
+			rx.map_err(|err| IoError::new(IoErrorKind::ConnectionAborted, err))
+		})
+		.into_future()
+		.and_then(|(_peer_id, kad_ctrl)| kad_ctrl);
+
+	// Note that we use a Box in order to speed compilation time.
+	Box::new(final_future) as Box<Future<Item = _, Error = _>>
+}
+
+/// Processes the information about a node.
+///
+/// - `original_addr` is the address used to originally dial this node.
+/// - `endpoint` is whether we dialed or listened to this node.
+/// - `transport` is used for the `nat_traversal` method.
+///
+/// Returns an error if the node has been refused access.
+fn process_identify_info(
+	shared: Arc<Shared>,
+	info: &IdentifyTransportOutcome,
+	original_addr: Multiaddr,
+	endpoint: Endpoint,
+	transport: &impl Transport
```

Comment:
please just use a generic parameter. IMO `impl Trait` should only be used where we can't name the type.

Reply:
But the function doesn't know whether the caller can name the type :confused: 
I would argue that using `impl Trait` in parameters is fine for private functions, because we can always change their signature later without breaking much.

Reply:
> But the function doesn't know whether the caller can name the type

I don't get it. If I write `fn foo<F: FnOnce>(f: F)` then I can call that with an unnameable type. It doesn't matter if the caller can name it.

Reply:
agree with @tomaka

the `impl` syntax is more concise - i'd prefer it over the clumsier generic syntax unless there's a reason to believe that it'll be called with explicit type args. in this case, it's private so there's no reason to believe that.